docs/decisions/0003-layered-cookiecutter.rst

```diff
@@ -0,0 +1,38 @@
+1. Layered Cookiecutters
+========================
+
+Status
+------
+
+Accepted
+
+Context
+-------
+
+We duplicate boilerplate files accross our multiple cookiecutters. Historically, its been difficult keeping the same files accross the multiple cookiecutters uptodate. We've decided to try layer cookiecutters. Each layer would have all the basic files that are need for that level of abstraction.
+
+Our approach splits our cookiecutters into two categorites: template, final_use. The template cookiecutters have all the base files, but they do not result in a workable repository output. They need to be used by final_use cookiecutters. The final_use cookiecutters are meant for final output and should result in a working directory for the users.
+
+Clarification: if a subsequent layer needs to modify a file that was created by previous layers, it will need to overwrite the file completely. Changing only segments of a file are thought to be outside the scope of this current work(though doing something like this is recommended in the future).
```

Comment:
1. Question: Is there a one-to-one relationship between templates and files? Or, even if a file is made from multiple templates, we only allow you to overwrite the entire file in the final-output cookiecutter.
2. Maybe?:
```
Note: The initial implementation only allows layers to use files created by a previous layer as-is, or overwrite it completely.
```

Reply:
A file can exist in multiple layers, but the final output will only have the file from the highest layer. For layering structure: [final-output]<-[template4]<-[template3]<-[template2]<-[template1], if a file exists in template{3,2,1}, only the version of the file in template3 will be outputted. so files from template2 and template1 will be over written.

Reply:
I'm wondering if this power is worth the complexity.  Feels a lot like the mess that configuration got into.  It may make it really difficult to understand what cookiecutters will actually get any new boiler-plate change that is added.

Reply:
@feanil: I'd like your thoughts, given your experience with configuration.

Reply:
I think the python template should definitely exists, I have been tempted to remove the django template, since it doesn't have alot in it.

Reply:
I decided to remove the django-template based on this convo and my own misgivings about the unnecessary complexity, I'm doing that in this PR: https://github.com/edx/edx-cookiecutters/pull/11

Reply:
I would recommend that where possible we don't build structure where you need to look at all the layers to reason about what the final output will be.  I think doing a full override is actually the better design, because it allows you to be clear: "If it's in the template it's the only file you have to look at to determine output."

I recommend that we explicitly call out that we *DO NOT* want to allow partial file updates because of the added complexity it generates.

Reply:
Thanks @feanil.

Reply:
okay, I think the current design fits with what you guys are saying. It does full override and I've changed the doc to remove any mentions of partial overwrites. I've also changed things such that there is only one layer available(there is no way to disallow more layers, I'll comment to suggest multiple layers are not recommended)